helm/flink-operator/templates/ingress.yaml

```diff
@@ -17,6 +17,23 @@
 ################################################################################
 ---
 {{- if .Values.ingress.create }}
+{{- if .Values.watchNamespaces }}
```

Comment:
How could we create the ingress when `watchNamespaces` is not configured? By default, the operator is watching all the namespaces, I am afraid it still could not update the ingress rules when running in a different namespace with operator.

All in all, I am suggesting to create the ingress in the java code, not the helm chart.

Reply:
You're right the all namespace scenario is not covered properly. We discussed this yesterday with @gyfora and @mbalassi and haven't had the chance to update the PR yet.

My main motivation behind using the Helm chart for creating the Ingresses was to use a more predictable/static manner. To my best knowledge Ingresses require at least an LB address, and a wildcard DNS entry with the current implementation to work properly, sometimes firewall/NW changes. 

So my suggestion for the ingress logic is the following:

- pre-create Ingresses for all watchNamespaces when the Operator is started (Let's use Java then)
- create Ingresses on the fly whenever a CR is created in a new namespace
- Set the operator deployment as the owner of all Ingresses and never delete them to avoid dynamic LB generation.

Let me know what you think @wangyang0918

Reply:
Thanks for sharing the progress. For the owner reference, I am afraid we could not set the owner of all the ingress in different namespaces to the operator deployment. Refer to here[1] for more information.

I am just thinking in a different way. Could we only create one ingress for the operator and let the operator do the HTTP requests redirecting? Benefit from this, we could easily integrate a unified authentication if necessary. I believe hadoop YARN[2] is running in a similar way to proxy the application master UI. 

And maybe we could also consider to provide a webUI for the operator itself in the future, which could greatly lower the bar to use our k8s operator.


[1]. https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference
[2]. https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/WebApplicationProxy.html#:~:text=The%20Web%20Application%20Proxy%20is,web%20based%20attacks%20through%20YARN.

Reply:
This should be doable with an ingress controller too, then. I'll dig into this.

Reply:
@wangyang0918 So it seems my assumption about the one LB address per Ingress is not correct:

> With the NGINX Ingress controller you can also have multiple ingress objects for multiple environments or namespaces with the same network load balancer; with the ALB, each ingress object requires a new load balancer.

![](https://d2908q01vomqb2.cloudfront.net/ca3512f4dfa95a03169c5a670a4c91a19b3077b4/2022/01/27/image7-3.png)

https://aws.amazon.com/blogs/opensource/network-load-balancer-nginx-ingress-controller-eks/

It seems we can create one ingress per deployment, and it get's merged by nginx. I'll test this and see if it works. Having a single ingress per deplyment would solve the ownership issue as well.

Reply:
It makes sense to me if the implementation does not rely on a specific cloud provision.

Reply:
I will get to the PR soon and verify it on the ACK(alibaba container kubernetes).